### PR TITLE
Make SDPA backend checks safe for unbacked head dims

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184983

`torch.export` can trace models that build repeat counts as tensors and
pass them to `repeat_interleave`. Without an explicit output size,
`repeat_interleave` creates unbacked symbolic sizes for the repeated K/V
head dimension. The fused SDPA backend selectors then used normal
`SymInt` comparisons such as `query_size_last == key_size_last`,
`head_dim <= max`, and alignment tests while deciding whether flash,
mem-efficient, or cuDNN kernels are viable. Those bool conversions guard
on the unbacked expressions before SDPA can fall back to math, producing
`GuardOnDataDependentSymNode` errors like `Eq(128, u0)`.

Switch those eligibility predicates to `sym_*` expressions evaluated with
`TORCH_GUARD_OR_FALSE`. Concrete and guardable symbolic inputs keep the
same result, while unbacked or unknown predicates conservatively reject
the fused backend and let the existing math fallback handle export. This
keeps the fix in backend selection instead of special-casing
`repeat_interleave` or SDPA export.

The main alternative would be requiring users to add `torch._check(...)`
for these repeated dimensions, but the backend picker only needs to know
whether a fused kernel is definitely legal. Treating unknown fused-kernel
preconditions as not viable is the narrower framework fix.

Fixes #170127
Generated by my agent

Test Plan:
- PYTHONPATH=/data/users/jansel/pytorch-issue-fixer/pytorch ninja -C build
- Minimized CPU repro for torch.export strict=False after repeat_interleave: passed
- Minimized CUDA/BF16 repro for torch.export strict=False after repeat_interleave: passed
- python test/test_transformers.py -k test_sdpa_export_repeat_interleave_unbacked_head_dim -v
- python test/test_transformers.py -k test_sdpa_export_unbacked_attn_mask -v
- python test/test_transformers.py -k test_fused_sdp_choice_cpu -v
- lintrunner -a
- git diff --cached --check

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR, resolving the ghstack mergeability conflict and rerunning CI. I also addressed the review feedback by converting the remaining direct SymInt comparisons and rewriting the SM86/SM89/SM120/SM121 FlashAttention training restriction as a known-supported predicate so uncertain head dims conservatively fall back.

**Comment by jansel** [comment]
My agent says the only remaining CI issue was an unclassified linux-jammy-py3.14t-clang18 Dynamo shard failure where run_test.py was killed with exit code 137, rather than a source-level test assertion. I reran the failed job and will revisit after CI reports the retry result.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resolved the SDPA selector conflict by preserving the newer ROCm AOTriton and cuDNN D256 rules while making their symbolic eligibility checks guard-safe. A full build, the CPU/CUDA export regressions, adjacent SDPA tests, and lintrunner all pass.

**Review by drisspg — APPROVED** [review]
Sure

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main to resolve the mergebot cherry-pick conflict, preserved the newer FA4, ROCm AOTriton, and cuDNN D192/D256 selector rules, and addressed the automated review by making the earlier zero-length and stride gates guard-safe and strengthening the regression to prove an unbacked repeated head dimension is exercised. The full build, targeted CPU/CUDA tests, and lintrunner pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the latest automated review by making the earlier batch, num-head, GQA, mask, and XPU selector checks guard-safe; extending GQA math preprocessing with deferred runtime assertions; and replacing the weak backend assertion with forced-fused export regressions that prove math fallback. I also fixed the SM86/SM120 coverage gate and added CUDA/XPU causal guard coverage. The full build, targeted CPU/CUDA export and GQA tests, and lintrunner all pass. The prior ROCm test_custom_ops failure was an unrelated timeout per Dr.CI; this update has triggered fresh CI.

**Comment by jansel** [comment]
@claude review these changes
